### ISSUE TITLE
"parameter hides field" inspection is not enabled 

### ISSUE THREAD

**Issue by lestephane** [gantsign/inspection-profile-intellij#1]
Enabling it would force the developer to prefix the field access with 'this.'  when a parameter and a field have the same name, while not forcing the developer to use 'this.' whenever a field is accessed. 

# Before:
```java
    private String value;
    public void append(String value) {
        value += value;
    }
```
# After:
![parameter_hides_field](https://user-images.githubusercontent.com/5460884/49802820-bb86f980-fd56-11e8-9dc3-9fc09c16fc5c.png)

```java
    private String value;
    public void append(String value) {
        this.value += value;
    }
```

What i like about the inspection is that, as long as the IDE reminds me of these hiding parameters, I can keep my code mostly free of **this.**, helping a bit with verbosity.

What is your opinion?

**Comment by freemanjp** [comment]
@lestephane, thank you for your suggestion. This is actually something I considered when creating this inspection profile but ultimately rejected. Let me explain my reasoning:

Your statement incorrect:
> Enabling it would force the developer to prefix the field access with 'this.' when a parameter and a field have the same name, while not forcing the developer to use 'this.' whenever a field is accessed.

Adding `this.` doesn't not resolve the "parameter hides field" warning. The only resolution is to use a different name for the field/parameter (or suppress the inspection).

While I appreciate the intent of this inspection, I don't like the effect this has on the code.

Take a Person bean:

```java
class Person {
    private String firstName;
    private String lastName;

    public Person(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }
}
```

The problem is how do I name my parameters/fields so there is no conflict.

### Solution 1: Use generic parameter name

For single parameter methods I could change the parameter name to something generic like `value` or `param` e.g.:

```java
class Person {
    private String firstName;
    private String lastName;

    public Person(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String value) {
        firstName = value;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String value) {
        lastName = value;
    }
}
```

This obviously doesn't help where there is more than one parameter (like the constructor). Also, if the method is long you may lose track of what the generic name is referring to (which wouldn't be a problem with a meaningful name).

### Solution 2: Use parameter name suffix

```java
class Person {
    private String firstName;
    private String lastName;

    public Person(String firstNameParam, String lastNameParam) {
        firstName = firstNameParam;
        lastName = lastNameParam;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstNameParam) {
        firstName = firstNameParam;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastNameParam) {
        lastName = lastNameParam;
    }
}
```

This works for every case, but it's quite verbose. My biggest complaint is this leaks into the API documentation. Also, do you add the suffix to every parameter or only where it hides a field and put up with having inconsistent parameter naming.

### Solution 3: Use field name prefix

```java
class Person {
    private String _firstName;
    private String _lastName;

    public Person(String firstName, String lastName) {
        _firstName = firstName;
        _lastName = lastName;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        _firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        _lastName = lastName;
    }
}
```

This is very similar to using `this.`; it's a bit more concise. It breaks standard Java conventions for field naming. It can also cause problems with code generation (when the code is generated using the field names).

### This inspection doesn't actually prevent bugs

The fact is it's just as easy to get the field/parameter names mixed up as it is with `this.`. All these solutions are just non standard conventions that achieve the same thing as `this.`.

### It's largely redundant

The fact is the IDE has different syntax highlighting to help you distinguish fields from parameters to help prevent mistakes. The IntelliJ IDEA will also detect most of the common cases where you get fields & parameters mixed up: e.g. assigning the parameter to itself `firstName = firstName`. I recommend using [Google Error Prone](https://github.com/google/error-prone) to catch these as part of your build as well.

### In conclusion

This is one of those ideas that looks good at first glance but on deeper inspection causes more problems than it solves.

It's also a problem that largely goes away if you use Kotlin due to it's support for properties, primary constructor properties, and extension functions.

**Comment by lestephane** [comment]
I can see you've put a lot of thought into all the inspections you've
activated. I'll think some more about this.

Do you configure Error Prone using an IDE config template, or using the
maven pom.xml?

**Comment by freemanjp** [comment]
Sorry, for the late reply. I use Error Prone in my Maven POM. I have a collection of Maven Archetypes that demonstrate my typical Maven config: [https://github.com/gantsign/maven-archetypes](https://github.com/gantsign/maven-archetypes).
